features/filesystem/i2ceeprom/I2CEepromBlockDevice.cpp

```diff
@@ -0,0 +1,138 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "I2CEepromBlockDevice.h"
+
+#define I2CEEPROM_TIMEOUT 10000
+ 
+
+I2CEepromBlockDevice::I2CEepromBlockDevice(PinName sda, PinName scl, uint8_t address,
+        bd_size_t size, bd_size_t block, int bus_speed)
+    : _i2c(sda, scl), _i2c_address(address)
+    , _size(size), _block(block)
+{
+    _i2c.frequency(bus_speed);
+}
+
+bd_error_t I2CEepromBlockDevice::init()
+{
+    return _sync();
+}
+
+bd_error_t I2CEepromBlockDevice::deinit()
+{
+    return 0;
+}
+
+bd_error_t I2CEepromBlockDevice::read(void *buffer, bd_addr_t addr, bd_size_t size)
+{
+    // Check the address and size fit onto the chip.
+    if (!is_valid_read(addr, size)) {
+        return BD_ERROR_PARAMETER;
+    }
+
+    _i2c.start();
+    if (!_i2c.write(_i2c_address | 0) ||
+        !_i2c.write((char)(addr >> 8)) ||
+        !_i2c.write((char)(addr & 0xff))) {
+        return BD_ERROR_DEVICE_ERROR;
+    }
+    _i2c.stop();
+
+    if (_i2c.read(_i2c_address, static_cast<char*>(buffer), size) < 0) {
+        return BD_ERROR_DEVICE_ERROR;
+    }
+
+    return 0;
+}
+ 
+bd_error_t I2CEepromBlockDevice::program(const void *buffer, bd_addr_t addr, bd_size_t size)
+{
+    // Check the address and size fit onto the chip.
+    if (!is_valid_program(addr, size)) {
+        return BD_ERROR_PARAMETER;
+    }
+        
+    // While we have some more data to write.
+    while (size > 0) {
+        _i2c.start();
+        if (!_i2c.write(_i2c_address | 0) ||
+            !_i2c.write((char)(addr >> 8)) ||
+            !_i2c.write((char)(addr & 0xff))) {
+            return BD_ERROR_DEVICE_ERROR;
+        }
+
+        for (bd_size_t i = 0; i < _block; i++) {
+            _i2c.write(static_cast<const char*>(buffer)[i]);
+        }
+        _i2c.stop();
+
+        bd_error_t err = _sync();
+        if (err) {
+            return err;
+        }
+
+        addr += _block;
+        size -= _block;
+        buffer = static_cast<const char*>(buffer) + _block;
+    }
+
+    return 0;
+}
+
+bd_error_t I2CEepromBlockDevice::erase(bd_addr_t addr, bd_size_t size)
+{
+    // No erase needed
+    return 0;
+}
+
+bd_error_t I2CEepromBlockDevice::_sync()
+{
+    // The chip doesn't ACK while writing to the actual EEPROM
+    // so loop trying to do a zero byte write until it is ACKed
+    // by the chip.
+    for (int i = 0; i < I2CEEPROM_TIMEOUT; i++) {
+        _i2c.start();
+        int ack = _i2c.write(_i2c_address | 0);
+        _i2c.stop();
```

Comment:
_i2c.write(_i2c_address | 0, 0, 0) could be used instead here.
In case of start / write(byte), there is no way for the driver to know whether a another byte will be written next. 
With the write(address, buf, length) API, this is clear that this is a check for availability of the device (length is 0).
We need this info for some of the stm32 device to have the check working ok..